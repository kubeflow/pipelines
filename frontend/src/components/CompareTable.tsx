/*
 * Copyright 2018 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { logger } from '../lib/Utils';
import { stylesheet, classes } from 'typestyle';
import { color } from '../Css';

const borderStyle = `1px solid ${color.divider}`;

const css = stylesheet({
  cell: {
    border: borderStyle,
    borderCollapse: 'collapse',
    padding: 5,
  },
  labelCell: {
    backgroundColor: color.lightGrey,
    fontWeight: 'bold',
    maxWidth: 200,
    minWidth: 50,
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
  },
  root: {
    border: borderStyle,
    borderCollapse: 'collapse',
  },
  row: {
    $nest: {
      '&:hover': {
        backgroundColor: '#f7f7f7',
      },
    },
  },
});

export interface CompareTableProps {
  rows: string[][];
  xLabels: string[];
  yLabels: string[];
}

class CompareTable extends React.PureComponent<CompareTableProps> {
  public render(): JSX.Element | null {
    const { rows, xLabels, yLabels } = this.props;
    if (rows.length !== yLabels.length) {
      logger.error(
        `Number of rows (${rows.length}) should match the number of Y labels (${yLabels.length}).`,
      );
    }
    if (!rows || rows.length === 0) {
      return null;
    }

    return (
      <table className={css.root}>
        <tbody>
          <tr className={css.row}>
            <td className={css.labelCell} />
            {/* X labels row */}
            {xLabels.map((label, i) => (
              <td key={i} className={classes(css.cell, css.labelCell)} title={label}>
                {label}
              </td>
            ))}
          </tr>
          {rows.map((row, i) => (
            <tr key={i} className={css.row}>
              {/* Y label */}
              <td className={classes(css.cell, css.labelCell)} title={yLabels[i]}>
                {yLabels[i]}
              </td>

              {/* Row cells */}
              {row.map((cell, j) => (
                <td key={j} className={css.cell} title={cell}>
                  {cell}
                </td>
              ))}
            </tr>
          ))}
        </tbody>
      </table>
    );
  }
}

export default CompareTable;
