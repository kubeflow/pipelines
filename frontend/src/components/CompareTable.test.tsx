/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { shallow } from 'enzyme';
import CompareTable from './CompareTable';

// tslint:disable-next-line:no-console
const consoleErrorBackup = console.error;
let consoleSpy: jest.Mock;

describe('CompareTable', () => {
  beforeAll(() => {
    consoleSpy = jest.spyOn(console, 'error').mockImplementation(() => null);
  });

  afterAll(() => {
    // tslint:disable-next-line:no-console
    console.error = consoleErrorBackup;
  });

  it('renders no data', () => {
    const tree = shallow(<CompareTable rows={[]} xLabels={[]} yLabels={[]} />);
    expect(tree).toMatchSnapshot();
  });

  const rows = [
    ['1', '2', '3'],
    ['4', '5', '6'],
    ['cell7', 'cell8', 'cell9'],
  ];
  const xLabels = ['col1', 'col2', 'col3'];
  const yLabels = ['row1', 'row2', 'row3'];

  it('logs error if ylabels and rows have different lengths', () => {
    shallow(<CompareTable rows={[rows[0], rows[1]]} xLabels={xLabels} yLabels={yLabels} />);
    expect(consoleSpy).toHaveBeenCalledWith(
      'Number of rows (2) should match the number of Y labels (3).',
    );
  });

  it('renders one row with three columns', () => {
    const tree = shallow(<CompareTable rows={rows} xLabels={xLabels} yLabels={yLabels} />);
    expect(tree).toMatchSnapshot();
  });
});
