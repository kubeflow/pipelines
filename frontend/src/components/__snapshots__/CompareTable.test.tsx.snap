// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompareTable renders no data 1`] = `""`;

exports[`CompareTable renders one row with three columns 1`] = `
<table
  className="root"
>
  <tbody>
    <tr
      className="row"
    >
      <td
        className="labelCell"
      />
      <td
        className="cell labelCell"
        key="0"
        title="col1"
      >
        col1
      </td>
      <td
        className="cell labelCell"
        key="1"
        title="col2"
      >
        col2
      </td>
      <td
        className="cell labelCell"
        key="2"
        title="col3"
      >
        col3
      </td>
    </tr>
    <tr
      className="row"
      key="0"
    >
      <td
        className="cell labelCell"
        title="row1"
      >
        row1
      </td>
      <td
        className="cell"
        key="0"
        title="1"
      >
        1
      </td>
      <td
        className="cell"
        key="1"
        title="2"
      >
        2
      </td>
      <td
        className="cell"
        key="2"
        title="3"
      >
        3
      </td>
    </tr>
    <tr
      className="row"
      key="1"
    >
      <td
        className="cell labelCell"
        title="row2"
      >
        row2
      </td>
      <td
        className="cell"
        key="0"
        title="4"
      >
        4
      </td>
      <td
        className="cell"
        key="1"
        title="5"
      >
        5
      </td>
      <td
        className="cell"
        key="2"
        title="6"
      >
        6
      </td>
    </tr>
    <tr
      className="row"
      key="2"
    >
      <td
        className="cell labelCell"
        title="row3"
      >
        row3
      </td>
      <td
        className="cell"
        key="0"
        title="cell7"
      >
        cell7
      </td>
      <td
        className="cell"
        key="1"
        title="cell8"
      >
        cell8
      </td>
      <td
        className="cell"
        key="2"
        title="cell9"
      >
        cell9
      </td>
    </tr>
  </tbody>
</table>
`;
